### ISSUE TITLE
STEllAR-GROUP/Vc slower than VCv1 possibly due to wrong instructions generated

### ISSUE THREAD

**Issue by DavidPfander-UniStuttgart** [STEllAR-GROUP/hpx#2702]
I'm observing significantly lower performance with VCv2 than with VCv1, performance of OctoTiger is approximately halved. To me it looks like the wrong instructions are generated as I'm seeing too many scalar mov ops (vmovsd instructions). I enabled AVX1/2 and FMA.

Possible reason for the observed problem might be that I'm compiling on a AMD Zen machine. I had to add a few lines to make Vc's cmake recognize it. That is in the following path.

```
From 9713208e013c895bc20e02fc9c6b9d583a31608f Mon Sep 17 00:00:00 2001
From: fire <fire@element.com>
Date: Sun, 18 Jun 2017 01:47:33 +0200
Subject: [PATCH] Added support switches for AMD Zen (copied from haswell)

---
 cmake/OptimizeForArchitecture.cmake | 13 ++++++++++---
 1 file changed, 10 insertions(+), 3 deletions(-)

diff --git a/cmake/OptimizeForArchitecture.cmake b/cmake/OptimizeForArchitecture.cmake
index 594dfd54..63724a97 100644
--- a/cmake/OptimizeForArchitecture.cmake
+++ b/cmake/OptimizeForArchitecture.cmake
@@ -152,8 +152,10 @@ macro(OFA_AutodetectX86)
             list(APPEND _available_vector_units_list "sse" "sse2" "sse3")
          endif(_cpu_model GREATER 2)
       endif(_cpu_family EQUAL 6)
-   elseif(_vendor_id STREQUAL "AuthenticAMD")
-      if(_cpu_family EQUAL 22) # 16h
+    elseif(_vendor_id STREQUAL "AuthenticAMD")
+      if(_cpu_family EQUAL 23) # 17h
+         set(TARGET_ARCHITECTURE "zen")
+      elseif(_cpu_family EQUAL 22) # 16h
          set(TARGET_ARCHITECTURE "AMD 16h")
       elseif(_cpu_family EQUAL 21) # 15h
          if(_cpu_model LESS 2)
@@ -315,6 +317,11 @@ macro(OFA_HandleX86Options)
    elseif(TARGET_ARCHITECTURE STREQUAL "AMD 14h")
       list(APPEND _march_flag_list "btver1")
       list(APPEND _available_vector_units_list "sse" "sse2" "sse3" "ssse3" "sse4a")
+   elseif(TARGET_ARCHITECTURE STREQUAL "zen")
+      list(APPEND _march_flag_list "znver1")
+      list(APPEND _available_vector_units_list "sse" "sse2" "sse3" "ssse3" "sse4.1" "sse4.2" "avx")
+      list(APPEND _available_vector_units_list "rdrnd" "f16c")
+      list(APPEND _available_vector_units_list "avx2" "fma" "bmi" "bmi2")
    elseif(TARGET_ARCHITECTURE STREQUAL "piledriver")
       list(APPEND _march_flag_list "bdver2")
       list(APPEND _march_flag_list "bdver1")
@@ -546,7 +553,7 @@ Other supported values are: \"none\", \"generic\", \"core\", \"merom\" (65nm Cor
 \"haswell\", \"broadwell\", \"skylake\", \"skylake-xeon\", \"cannonlake\", \"silvermont\", \
 \"goldmont\", \"knl\" (Knights Landing), \"atom\", \"k8\", \"k8-sse3\", \"barcelona\", \
 \"istanbul\", \"magny-cours\", \"bulldozer\", \"interlagos\", \"piledriver\", \
-\"AMD 14h\", \"AMD 16h\".")
+\"AMD 14h\", \"AMD 16h\", \"AMD 17h\".")
    elseif("${CMAKE_SYSTEM_PROCESSOR}" MATCHES "(arm|aarch32|aarch64)")
       set(TARGET_ARCHITECTURE "auto" CACHE STRING "CPU architecture to optimize for. \
 Using an incorrect setting here can result in crashes of the resulting binary because of invalid instructions used. \
-- 
2.11.0
```

**Comment by hkaiser** [comment]
@mattkretz adding you to this discussion...

**Comment by mattkretz** [comment]
I just pushed my work on vectorizing the fixed_size ABI to mkretz/datapar. It also contains a similar patch to recognize AMD Zen, so that should work out of the box now. Also, I added a few always inline attributes that I found missing. It would be helpful if you can retest with those changes.

You're saying that on the same machine, using Vc 1.3, you get approximately twice the performance as with Vc 2?

I really have no idea at this point how this would happen other than the use of fixed_size / SimdArray. Updating would fix it then.

**Comment by DavidPfander-UniStuttgart** [comment]
Ok, I'll try https://github.com/VcDevel/Vc/tree/mkretz/datapar then. If that branch is the reference for datapar 2, maybe we should get rid of the https://github.com/STEllAR-GROUP/Vc.

**Comment by hkaiser** [comment]
@DavidPfander-UniStuttgart This is the desired solution, eventually. However, I think there are things in that repo which @brycelelbach was planning to push upstream. Matthias, do you have any recollection of what Bryce did there?

**Comment by mattkretz** [comment]
I'll look into rebasing wash's branch onto mine. It's some compatibility work and a few math features that he added IIRC.

**Comment by mattkretz** [comment]
Merged in mkretz/abs_sqrt branch now. After CI gives a green light, I'll merge it back to mkretz/datapar.

**Comment by DavidPfander-UniStuttgart** [comment]
Thanks you very much, Matthias. This is extremely helpful for our KNL work!

**Comment by mattkretz** [comment]
The link is hard to follow, it's href points to nirvana.
Anyway, the picture is hard to interpret. Perfect for fixing this would be a minimal reduced testcase (you could give it a shot using multidelta and some smart "grep" on the generated asm).
Else, better give me the binary itself and I can disassemble and look. Or maybe I can reproduce the build here? Let me know how.

**Comment by mattkretz** [comment]
hmm, so I think I can guess the `callq` now. The compiler decided not to inline the implementation of `fixed_size_datapar<double, 16> + fixed_size_datapar<double, 16>`. That's a crazy choice, but one I've seen way too often. :-( Maybe I have to put a `gnu::always_inline` on it, even though in some cases (e.g. <double, 32> without AVX) the function can become a bit largish. In any case, the choice of const-ref arguments on those `fixed_size_datapar_impl` functions is wrong. I'll come up with something and let you know.

**Comment by DavidPfander-UniStuttgart** [comment]
Should I open a issue in VcDevel/Vc about this?

And sorry for the messed up link. 

The source code [here](https://github.com/STEllAR-GROUP/octotiger/blob/double_loop_comp_int/src/taylor_set_basis.hpp) shows that this specific piece code is vector memory accesses and vector arithmetic only, it's also branch free. There should only be very few scalar instruction for the constant indexing.

I'll also add gcc assembly for this in a few minutes.

**Comment by DavidPfander-UniStuttgart** [comment]
Here is a vtune exported assembly listing for the whole method.

[vtune_assembly_export.txt](https://github.com/STEllAR-GROUP/hpx/files/1094167/vtune_assembly_export.txt)

**Comment by DavidPfander-UniStuttgart** [comment]
Build scripts for octotiger. Uncompless to an arbitrary folder, setup the paths in the bash file and run it. (Derived from our Circle CI scripts). It already checks out branches I'm using. For OctoTiger, this is my current development branch.
[octotiger_build_scripts_ubuntu.zip](https://github.com/STEllAR-GROUP/hpx/files/1094575/octotiger_build_scripts_ubuntu.zip)

If you want to verify your build, run
```
./build/octotiger -t1 -Problem=moving_star -Max_level=3 -Odt=0.3 -Stoptime=0.2 -Xscale=20.0 -Omega=0.1  -Stopstep=9
```
The final lines of the output (after ~1 min) should be
```
L1, L2
             rho 2.751004e-02 3.057842e-02
            egas 3.581150e-02 4.770200e-02
              sx 5.006693e-02 5.305132e-02
              sy 4.844836e-02 6.009045e-02
             tau 2.426892e-02 2.627559e-02
    primary_core 2.751004e-02 3.057842e-02
Outputing...
Done...
...:
   Total: 71.583
   Computation: 67.7001
   Regrid: 0.0727124
   Compare Analytic: 0.185826
   Find Localities: 0
total_neighbors: 18144
missing_neighbors: 12768
Exiting...
```

**Comment by hkaiser** [comment]
@DavidPfander-UniStuttgart, @mattkretz : this is kind of unrelated to this repository. Would you mind moving this to the Vc repo instead?
